storage/namespace_map_gen.go

```diff
@@ -0,0 +1,259 @@
+
+// databaseNamespacesMap uses the genny package to provide a generic hash map that can be specialized
+// by running the following command from this root of the repository:
+// ```
+// make hashmap-gen pkg=outpkg key_type=Type value_type=Type out_dir=/tmp
+// ```
+// Or if you would like to use bytes or ident.ID as keys you can use the
+// partially specialized maps to generate your own maps as well:
+// ```
+// make byteshashmap-gen pkg=outpkg value_type=Type out_dir=/tmp
+// make idhashmap-gen pkg=outpkg value_type=Type out_dir=/tmp
+// ```
+// This will output to stdout the generated source file to use for your map.
+// It uses linear probing by incrementing the number of the hash created when
+// hashing the identifier if there is a collision.
+// databaseNamespacesMap is a value type and not an interface to allow for less painful
+// upgrades when adding/removing methods, it is not likely to need mocking so
+// an interface would not be super useful either.
+type databaseNamespacesMap struct {
+	_databaseNamespacesMapOptions
+
+	// lookup uses hash of the identifier for the key and the MapEntry value
+	// wraps the value type and the key (used to ensure lookup is correct
+	// when dealing with collisions), we use uint64 for the hash partially
+	// because lookups of maps with uint64 keys has a fast path for Go.
+	lookup map[databaseNamespacesMapHash]databaseNamespacesMapEntry
+}
+
+// _databaseNamespacesMapOptions is a set of options used when creating an identifier map, it is kept
+// private so that implementers of the generated map can specify their own options
+// that partially fulfill these options.
+type _databaseNamespacesMapOptions struct {
+	// hash is the hash function to execute when hashing a key.
+	hash databaseNamespacesMapHashFn
+	// equals is the equals key function to execute when detecting equality.
+	equals databaseNamespacesMapEqualsFn
+	// copy is the copy key function to execute when copying the key.
+	copy databaseNamespacesMapCopyFn
+	// finalize is the finalize key function to execute when finished with a
+	// key, this is optional to specify.
+	finalize databaseNamespacesMapFinalizeFn
+	// initialSize is the initial size for the map, use zero to use Go's std map
+	// initial size and consequently is optional to specify.
+	initialSize int
+}
+
+// databaseNamespacesMapEntry is an entry in the map, this is public to support iterating
+// over the map using a native Go for loop.
+type databaseNamespacesMapEntry struct {
+	// key is used to check equality on lookups to resolve collisions
+	key _databaseNamespacesMapKey
+	// value type stored
+	value databaseNamespace
+}
+
+type _databaseNamespacesMapKey struct {
+	key      ident.ID
+	finalize bool
+}
+
+// Key returns the map entry key.
+func (e databaseNamespacesMapEntry) Key() ident.ID {
+	return e.key.key
+}
+
+// databaseNamespace returns the map entry value.
+func (e databaseNamespacesMapEntry) databaseNamespace() databaseNamespace {
```

Comment:
how about just calling this `Value()` instead? would make the usage of the generated map a little more standard - always have to use `Key()/Value()` instead of type specific method names to access